docs/source/advanced/optimized_multi_gpu.rst

```diff
@@ -0,0 +1,661 @@
+
+To use Sharded Training, you need to first install FairScale using the command below.
+
+.. code-block:: bash
+
+    pip install fairscale
+
+
+.. code-block:: python
+
+    # train using Sharded DDP
+    trainer = Trainer(plugins='ddp_sharded')
+
+Sharded Training can work across all DDP variants by adding the additional ``--plugins ddp_sharded`` flag.
+
+Internally we re-initialize your optimizers and shard them across your machines and processes. We handle all communication using PyTorch distributed, so no code changes are required.
+
+----------
+
+.. _fully-sharded:
+
+Fully Sharded Training
+^^^^^^^^^^^^^^^^^^^^^^
+
+.. note::
+    Fully Sharded Training is in beta and the API is subject to change. Please create an `issue <https://github.com/PyTorchLightning/pytorch-lightning/issues>`_ if you run into any issues.
+
+`Fully Sharded <https://fairscale.readthedocs.io/en/latest/api/nn/fsdp.html>`__ shards optimizer state, gradients and parameters across data parallel workers. This allows you to fit much larger models onto multiple GPUs into memory.
+
+By default, Fully Sharded acts similar to :ref:`sharded` which shards optimizer states and gradients. If you can train with default Fully Sharded, it is recommended to just use :ref:`sharded`.
+
+Shard Parameters to Reach 10+ Billion Parameters
+""""""""""""""""""""""""""""""""""""""""""""""""
+
+To reach larger parameter sizes and be memory efficient, we have to shard parameters. There are various ways to enable this.
+
+Auto Wrap
+"""""""""
+
+``auto_wrap`` will recursively wrap modules within the ``LightningModule`` with nested Fully Sharded Wrappers,
+signalling that we'd like to partition these modules across data parallel devices, discarding the full weights when not required (information `here <https://fairscale.readthedocs.io/en/latest/api/nn/fsdp_tips.html>`__).
+
+Enabling `auto_wrap` doesn't require code changes, however can have varying level of success based on the complexity of your model. **Auto Wrap does not support models with shared parameters**, use :ref:`manual-wrap` instead.
+
+.. code-block:: python
+
+    from pytorch_lightning import Trainer
+    from pytorch_lightning.plugins import DeepSpeedPlugin
+
+    class MyModel(pl.LightningModule):
+        ...
+
+    model = MyModel()
+    trainer = Trainer(gpus=4, plugins='ddp_fully_sharded_auto_wrap', precision=16)
+    trainer.fit(model)
+
+    trainer.test()
+    trainer.predict()
+
+
+.. _manual-wrap:
+
+Manual Wrap
+"""""""""""
+
+To activate parameter sharding, you can also wrap layers using provided ``wrap`` or ``auto_wrap`` functions as described below.
+
+When not using Fully Sharded these wrap functions are a no-op. This means once the changes have been made, there is no need to remove the changes for other plugins.
+
+This is a requirement for really large models and also saves on instantiation time as modules are sharded instantly, rather than after the entire model is created in memory.
```

Comment:
Thanks @min-xu-ai, could you go into more details as to what is different here compared to DDP?

Reply:
Sure. With DDP, you can have this:
```
rank 0               rank 1
m = model()    m = model()  <------ two ranks may have different weights due to different random seeds
train(m)            train(m)  <---------------- weights are synced by ddp
```

With FSDP, since `m` is sharded, parts of the weights will be from rank 0 and parts of the weights will be from rank 1 when sharding happens. That can break the weight init assumptions, like zero mean and unit stddev etc.

Therefore, until FSDP can sync weights between ranks, weight init needs to be very careful with FSDP.

Reply:
What would be a potential solution here? use torch.distributed communications to sync global stats across all shards when initializing the model? It would be good to have a solution in place in the docs for users to have an example!

Reply:
Just following up here because it might be a solution @min-xu-ai, but using `SummonFullParams` may be a way to init the model locally if the model would fit into memory, and then broadcast results. I'll add this into the FSDP docs in time when we merge the feature in!